cellarium/ml/transforms/transforms.py

```diff
@@ -1,108 +1,251 @@
 # Copyright Contributors to the Cellarium project.
 # SPDX-License-Identifier: BSD-3-Clause
 
+from functools import cache
+from typing import Any
+
+import numpy as np
 import torch
+from numpy.typing import ArrayLike
 from torch import nn
 
 
-class NormalizeTotal(nn.Module):
+class DivideByScale(nn.Module):
     """
-    Normalize total gene counts per cell to target count.
+    Divide gene counts by a scale.
+
+    .. math::
+
+        y_{ng} = \\frac{x_{ng}}{\\mathrm{scale}_g + \\mathrm{eps}}
 
     Args:
-        target_count:
-            Target gene epxression count.
+        scale_g:
+            A scale for each gene.
         eps:
             A value added to the denominator for numerical stability.
+        feature_schema:
+            The variable names schema for the input data validation.
+            If ``None``, no validation is performed.
     """
 
-    def __init__(
-        self,
-        target_count: int = 10_000,
-        eps: float = 1e-6,
-    ):
+    def __init__(self, scale_g: torch.Tensor, eps: float = 1e-6, feature_schema: ArrayLike | None = None) -> None:
         super().__init__()
-        self.target_count = target_count
+        self.scale_g: torch.Tensor
+        self.register_buffer("scale_g", scale_g)
         self.eps = eps
+        if feature_schema is not None:
+            feature_schema = np.array(feature_schema)
+        self.feature_schema = feature_schema
+
+    def forward(
+        self,
+        x_ng: torch.Tensor,
+        feature_g: np.ndarray | None = None,
+    ) -> torch.Tensor:
+        """
+        Args:
+            x_ng:
+                Gene counts.
+            feature_g:
+                The list of the variable names in the input data. If ``None``, no validation is performed.
+
+        Returns:
+            Gene counts divided by scale.
+        """
+        if self.feature_schema is not None and feature_g is not None:
```

Comment:
Let's cache the validations:
```
@cache
def input_features_are_valid(self, feature_g: np.ndarray) -> bool:
    return np.array_equal(self.feature_schema, feature_g)

def forward(...):
    ...
    assert x_ng.shape[1] == len(feature_g), "The number of x_ng columns must match the feature_g length."
    assert self.input_features_are_valid(feature_g), "feature_g must match the feature schema".
    ...
```

... and I'm hoping `hash(feature_g)` is faster than `np.array_equal(self.feature_schema, feature_g)` ... right?!

Reply:
Another consideration is that `np.ndarray` is not hashable and first needs to be converted to `tuple` (hashable). Then the question is is `hash(tuple(feature_g.tolist()))` is faster than `np.array_equal(self.feature_schema, feature_g)`?

Reply:
Maybe we can we `id(feature_g)` as a hash? Example of that in funsor: https://github.com/pyro-ppl/funsor/blob/c96e70c89aaa0bda8dc132ae1c3498a681f618b8/funsor/interpretations.py#L66

Reply:
Some tests:

```py
In [4]: import numpy as np

In [5]: x = np.array([f"gene_{i}" for i in range(30000)])

In [6]: x
Out[6]: 
array(['gene_0', 'gene_1', 'gene_2', ..., 'gene_29997', 'gene_29998',
       'gene_29999'], dtype='<U10')

In [7]: %timeit x.tolist()
1.23 ms ± 6.69 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [8]: %timeit tuple(x.tolist())
1.46 ms ± 10.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [9]: %timeit hash(tuple(x.tolist()))
2.14 ms ± 24.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: %timeit np.array_equal(x, x)
325 µs ± 2.87 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```